### ISSUE TITLE
Replicating the paper's 94.0 % accuracy on CIFAR10 (Table 1)

### ISSUE THREAD

**Issue by amirkhojaste** [HazyResearch/tanda#16]
Hi there and thanks for open sourcing the code, 

I am trying to replicate the results of Table 1 in the paper for CIFAR10. I have used the following script, but after 300 epochs the average accuracy was around 72% far away from 94% claim in the paper. Could you please advise on how to get the 94% accuracy of the Table 1 in paper? 

Reading the issues here I found that I need to run the TAN model for only 1 epoch. However, I have not found on how many epochs the end model needs to be run anywhere in the paper so I ran it for 300 epochs. 

The script is as below:

#!/bin/bash
source set_env.sh
python3 experiments/cifar10/train.py \
        --run_type tanda-full \
        --generator lstm \
        --gen_config init_type=train,feed_actions=True,n_stack=1,logit_range=2.0 \
        --n_per_class 5000 \
        --n_sample 5 \
        --transformer image \
        --gamma 0.5 \
        --gen_lr 0.0001 \
        --disc_lr 0.00001 \
        --mse_term 0.001 \
        --mse_layer 1 \
        --seq_len 10 \
        --per_img_std False \
        --batch_size 32 \
        --n_epochs 1 \
        --run_name cifar_tan_full \
        --is_test True \
        --end_epochs 300 \
        --p_transform_init 1.0 \
        --p_transform_init 1.0 \
        --p_transform_max 1.0 \
        --save_model True

**Comment by henryre** [comment]
Hi @amirkhojaste, thanks for trying out TANDA. Can you try training the TAN and end model separately with the following configs?

[TAN config](https://github.com/HazyResearch/tanda/blob/master/example-scripts/cifar-example.sh)
[End model config](https://github.com/HazyResearch/tanda/blob/master/experiments/cifar10/config/end_model_config.json)

**Comment by henryre** [comment]
In the end model config, you can increase n_per_class to e.g. 5000 to learn on the full dataset.

**Comment by amirkhojaste** [comment]
Hey @henryre, thanks for your prompt response! I will give it a shot and will get back if I have any questions. 

Regards,
Amir

**Comment by henryre** [comment]
Hi @amirkhojaste, we're double checking that the hyperparameters we put up are the ones we ended up using. We'll get back to you ASAP.

**Comment by amirkhojaste** [comment]
Thanks a lot for your response. I appreciate your time.

**Comment by henryre** [comment]
You can try out

```
{
    "name" : "cifar_full_scale",
    "parameters" : {
        "validation_set": [false],
        "transformer": ["image-pct"],
        "p_transform_init": [1.0],
        "p_transform_max": [1.0],
	"p_transform_drop": [140],
        "end_epochs" : [150],
        "end_per_img_std" : [true],
        "n_per_class" : [4000],
        "end_weight_decay" : [0.0002],
        "end_optimizer" : ["momentum"],
        "end_lr_mode" : ["schedule"],
        "end_lr_schedule": [1],
        "eval_every": [10],
        "subsample_seed": [1701],
        "seq_len": [10],
        "ls_term": [0.0]
    }
} 
```

You'll need to add the follow LR schedule to `train_scripts.py`

```
[(0, 0.1), (10000, 0.01)]
```

**Comment by ajratner** [comment]
@amirkhojaste thanks for your help in getting this resolved!!  Let us know how that one goes!

**Comment by amirkhojaste** [comment]
Thanks @henryre and @ajratner.  But I'm still at 92% accuracy.

**Comment by henryre** [comment]
Hey @amirkhojaste thanks for getting back to us! We're rerunning things on our end to make sure we have the correct hyperparameters. We'll get back to you when we find out what's going on.

**Comment by amirkhojaste** [comment]
Any updates? 

Thanks,
Amir

**Comment by henryre** [comment]
Hi @amirkhojaste, sorry for the delay and thanks a ton for your diligence on this. I was able to reproduce our results from the paper: 94.1% (original was 94.0%) on the full CIFAR-10 dataset after 150 epochs using the TAN training script in the repo (`example-scripts/cifar-example.sh`). Here were the material changes between my setup and the settings you tried before:

1. Using TensorFlow 1.2: I believe the key issue with using our code and a newer TF version is that batch norm will be applied incorrectly (as you highlighted in #18). This impacts both the TAN and end model. I'm opening an issue to fix forward compatibility, and updating the README and package requirements accordingly in the meantime (i.e. telling people to only use TF 1.2).
2. Incorrect learning rate schedule for the end model: we used the suggested learning rate schedule from the original ResNet paper, but we rescaled it for the 10% dataset experiments. This turns out to be really bad for the full dataset. I'm adding the full scale learning rate schedule to the repo, and will add a full-scale end model training config as well.

Again, thanks for your help with this. Let me know if you have any more questions!
